index.d.ts

```diff
@@ -0,0 +1,130 @@
+declare module 'fluture' {
+  interface ResolveFunction<T> {
+    (value: T): void
+  }
+  interface RejectFunction {
+    (error: Error): void
+  }
+
+  interface FunctionReturning<T extends Function> {
+    (...t: any[]): T
+  }
+  abstract class Future<T> {
+    done<T> (fn: (err: Error, value: T) => void): Future<T>
+  }
+
+  function after<T> (duration: number, value: string): Future<T>
+
+  function and<A, B> (a: Future<A>, b: Future<B>): Future<A | B>
+
+  function ap<S, T> (fn: Future<(t: T) => S>, src: Future<T>): Future<S>
+
+  function attempt<T> (f: () => T): Future<T>
+
+  function bimap<A, B, AA, BB> (left: (a: A) => AA,
+                                right: (a: B) => BB): Future<AA>
+
+  function both<A, B> (a: Future<A>, b: Future<B>): Future<[A, B]>
+
+  function cache<T> (f: Future<T>): Future<T>
+
+  function chain<T, S> (fn: (s: S) => Future<Future<T>>, src: Future<S>): Future<T>
+
+  function encase2<T, A, B> (fn: (a: A, b: B) => T, a: A, b: B): Future<T>
+
+  function encase3<T, A, B, C> (fn: (a: A, b: B, c: C) => T, a: A, b: B, c: C): Future<T>
+
+  function encase<T> (fn: FunctionReturning<T>, ...t: any[]): Future<T>
+  function encase<T> (fn: FunctionReturning<T>): {(...t: any[]): Future<T>}
+
+
+  // TODO: Need Help
+  function encaseN2<T> (): Future<T>
+
+
+  // TODO: Need Help
+  function encaseN3<T> (): Future<T>
+
+  // TODO: Need Help
+  function encaseN<T> (): Future<T>
+
+
+  // TODO: Need Help
+  function encaseP2<T> (): Future<T>
+
+  // TODO: Need Help
+  function encaseP3<T> (): Future<T>
+
+  // TODO: Need Help
+  function encaseP<T> (): Future<T>
+
+  // TODO: Need Help
+  function extractLeft<T> (): Future<T>
+
+  // TODO: Need Help
+  function extractRight<T> (): Future<T>
+
+  function fold<A, AA, B, BB> (a: (x: A) => AA,  b: (x: B) => BB): Future<BB|AA>
+
+  // TODO: Need Help
+  function fork<T> (): Future<T>
+
+  // TODO: Need Help
+  function go<T> (): Future<T>
+
+  // TODO: Need Help
+  function hook<T> (): Future<T>
+
+  function isFuture<T> (fut: any): boolean
+
+  // TODO: Need Help
+  function isNever<T> (): boolean
+
+  // TODO: Need Help
+  function lastly<T> (): Future<T>
+
+  // TODO: Need Help
+  function lastly<T> (): Future<T>
+
+  // TODO: Need Help
+  function map<T> (): Future<T>
+
+  // TODO: Need Help
+  function mapRej<T> (): Future<T>
+
+  // TODO: Need Help
+  function never<T> (): Future<T>
+
+  // TODO: Need Help
+  function node<T> (): Future<T>
+
+  function of<T> (value: T): Future<T>
+
+  function or<A, B> (a: Future<A>, b: Future<B>): Future<A | B>
```

Comment:
Do you think it would make more sense to have something like —
```ts
or<L1, R1, L2, R2> (a: Future<L1, R1>, b: Future<L2, R2>): Future<L1|L2, R1|R2>
```